### ISSUE TITLE
Native resource for fetching kubeconfig

### ISSUE THREAD

**Issue by worldofgeese** [ionos-cloud/terraform-provider-ionoscloud#66]
### Current Provider Version

```
6.0.0-beta.2
```

### Use-cases
I want to easily make use of the `helm_release` and `banzaicloud/k8s` resources without using `local-exec` to pull down a kubeconfig file. Local files are often untouchable in CI/CD pipelines and are otherwise non-deterministic, unreliable objects.

### Attempted Solutions
Here's what Famly uses now to get a usable kubeconfig:
```
data "ionoscloud_k8s_cluster" "yb1" {
  id = "${ionoscloud_k8s_cluster.yb1.id}"
}
resource "null_resource" "getcfg" {
  provisioner "local-exec" {
    command = <<EOT
      curl --include \
     --request GET \
     --user "${var.username}:${var.secret}" \
     https://api.ionos.com/cloudapi/v6/k8s/${ionoscloud_k8s_cluster.yb1.id}/kubeconfig | sed -n '17,34p' > kubeconfig.yaml && sleep 30
 EOT
  }
  depends_on = [ionoscloud_k8s_cluster.yb1]
  provisioner "local-exec" {
    when    = destroy
    command = <<EOT
       rm kubeconfig.yaml 
  EOT
  }
}
provider "helm" {
  kubernetes {
    config_path = "./kubeconfig.yaml"
  }
} 
resource "helm_release" "ingress-nginx" {
  repository       = "https://kubernetes.github.io/ingress-nginx"
  name             = "ingress-nginx"
  chart            = "ingress-nginx"
  namespace        = "ingress-nginx"
  create_namespace = true
}

```

### Proposal
It could be as simple as:
```
provider "helm" {
  kubernetes {
    config_path = data.ionoscloud_k8s_cluster.this.kube_config
  }
}
```
However, `helm_release` will error with a `error loading config file: ... file name too long`.

Here's an example of how to fetch a certificate for AWS EKS:
```
provider "helm" {
  kubernetes {
    host                   = data.aws_eks_cluster.cluster.endpoint
    cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
    exec {
      api_version = "client.authentication.k8s.io/v1alpha1"
      args        = ["eks", "get-token", "--cluster-name", data.aws_eks_cluster.cluster.name]
      command     = "aws"
    }
  }
}
```

How the rancher provider offers it (in my opinion the most elegant):
```
provider "k8s" {
  host = rke_cluster.rancher_cluster.api_server_url

  client_certificate     = rke_cluster.rancher_cluster.client_cert
  client_key             = rke_cluster.rancher_cluster.client_key
  cluster_ca_certificate = rke_cluster.rancher_cluster.ca_crt

  load_config_file = false
}
```

**Comment by worldofgeese** [comment]
I wanted to follow up on this.

The `data.ionoscloud_k8s_cluster.this.kube_config` resource doesn't appear to output valid json or yaml:
```
{apiVersion:v1,kind:Config,preferences:{},current-context:cluster-admin@famly-prod,clusters:[{cluster:{certificate-authority-data:LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUZFVENDQXZtZ0F3SUJBZ0lVQ1VxdEJwOHNucU5WVFFQTWhJcXljNkE5UlhJd0RRWUpLb1pJaHZjTkFRRUwKQlFBd0ZURVRNQkVHQTFVRUF3d0thM1ZpWlhKdVpYUmxjekFlRncweU1UQTNNall4TkRBek5UUmFGdzB6TVRBMwpNalF4TkRBek5UUmFNQlV4RXpBUkJnTlZCQU1NQ210MVltVnlibVYwWlhNd2dnSWlNQTBHQ1NxR1NJYjNEUUVCCkFRVUFBNElDRHdBd2dnSUtBb0lDQVFEZWRQb2IyeDViQkd4MUdDUllPbkE1Y0VpNWVwemZSUENXZ2ZONDhRNHUKWlFKend0d1g0YXR1K3pVWUlxMkhURi9Ob01YTUNLMjUrWjBMdG5hNk9LUTI1V1B1SEJhMzJOVGZDWjI4dGJoSwpVbURkRUdGRlpsZHpQQXhpWUlJeHhZY1R5aWZxTTU1UHhJS2UyOXpQczFMWXRHVHlwcXZmQzVGNmRDNThZNGpFClAyUHJNYnJnY2l3VFNLb3liTG9kRG8rYkVVa2pMdEFFSHJzQkxLbCtOODlUb1JVRFNoMXJ2YWdaZXA1dXFDa3UKbnFoYlNDRExqMzZZWEQvZWg4RENvaFJIN1pseDNvZUkvN0kzVjViemFGd0trWHBjUjMwVkZGb1VBSU0wV29KUgpiZ3RGQTdQK3duajErc3Z6MFZmQm9XN1JIbXd2aXVLa3VJWjFtOFVPcktkSG1XYno1eDVTRmI2am1jYTlCVjlxClZISWQvTy8yRmk0UXJmTitJWXJNUllyT1hmYUdnUEsxQzhzSjBCakJDYjNSbWRXSjZ0ZXYvVU9nS1J0bnJOV24KTE9MclArWXQyTGFwM1BXSWs1bXBFaXErVWQvNUY2ZnBEbmJNZGl0SlYrQ2F2MEt1ZW9odjFhOVdSQ2YrWFh2VAo0c1RVUTBEODluWlNzMlYwR3pIaUZCTzlpUlRFMmRUUkFWNWNCYVBOVWdwVHE5eVhxeUEzQTN6NkxNTEE2MTVJCmFDTjBtT0Z5UUhEZWcyQ1d4NElYK2l4NjY0dEYxZmFHenlNd05EK1lvSnBxWjlpdTA2L0RjMW1HakxoYlk1M3IKd0dnVXUrdk1OdDZhaHhaNWtPeWl1aGo2YnhseVhmYzBqTFlkYVBWcHl0d2t2TGxGWmFHWmw5YUZOU1NWSFJTQwpTd0lEQVFBQm8xa3dWekFWQmdOVkhSRUVEakFNZ2dwcmRXSmxjbTVsZEdWek1BNEdBMVVkRHdFQi93UUVBd0lDCnBEQVBCZ05WSFJNQkFmOEVCVEFEQVFIL01CMEdBMVVkRGdRV0JCVHhLeEpkV1h4YjZ3QU82bk1YMEpCK2pLR0sKSGpBTkJna3Foa2lHOXcwQkFRc0ZBQU9DQWdFQWdZc1JhRmhmb0FmMS84K3RsVW56bFpXOGVYQjlKZDcxRHB6aQpINnpvU08wZXVzaitxSXhwczd5WVU1S3hlalh6S3VZWUp1aEtzR0ZJdXkxTVlZbldQbUpUWFVYNnU1T2QzbFVhClMxOWdLMDk4SWpnZlZNbEFkWnlqejY1c0VxYmpiWmVwSlFkVjV0elQwYU9zL21WRUN5bjllakZnNm14dUw1THkKRGpCc3FrL3U4NGdFbkQxUktFc1VnRFFXbFAyaHJxQllwYkF3ajBvbTFYNVoyRUIxWVF6MTlzVGt5RXBEZFdLbApJc2ppaCtWMUdZaDlEcjd5OXJEQXgxcEpNNU8vT2xxRjlWMTlHQ09hQTBuL0JoVjBjeTNaTlE3aUZDVDNrM2FyCnNpMlJ1b25ON1pIMGFBM3Z5clY3MjJRY3N5bmN5Ulc5MDVzWHBlSVVQNGgzRDBDcjJFUC95dTRwZlRyckdNT2YKWmJKRmZkSVByTitEcWxTOEF5aitzS3plQW8vN0xMOUNkN2FOYlpaQWgzMUxNYzZKd2xLQXoxYjAzNkowN1MwcgpUckxuU0JSanR4OE1wczduSG5naFg2RnM0YU5nOWZ4c01CTjhhejZiWEtaL0dLMFRRQnloT3A4R2I3RTMzK2dqCitoRFFOMGlQbUdHaFhsbXdXbm9WYjRoV240ZmptOXRSY1AxaFNOemJJR1JWbzlOUmVodGRPYk9DeWNyMWVjTWIKRWJFRG91dGNLY3huaXVXb2ZMUlRUcVhRb1o0dUQ4NWM4U0RjMUNCTlhQN1l0OXYzK2VNeDVkU0hkdWNweEw0YQpYL21JUU5kOVgxZnRmd2wxeDhaYVBlY25qWmJxdFhQR1hkWTBlajh1YkNqUjNuRmlrRUI0Q3hBRVNnaHIyUW9FCjluMDc5czQ9Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K,server:https://cp-18893.cluster.ionos.com:10466},name:famly-prod}],contexts:[{context:{cluster:famly-prod,user:cluster-admin},name:cluster-admin@famly-prod}],users:[{name:cluster-admin,user:{token:eyJhbGciOiJSUzI1NiIsImtpZCI6Imx1N3NUX3dFODVhRXBPYnJpSmZhTlBkZjRoZWlXVDNlSFAxY2FFeE1ubDQifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJrdWJlLXN5c3RlbSIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VjcmV0Lm5hbWUiOiJjbHVzdGVyLWFkbWluLXRva2VuLXRuemZxIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZXJ2aWNlLWFjY291bnQubmFtZSI6ImNsdXN0ZXItYWRtaW4iLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC51aWQiOiI0MTY2YmFhZC0yMWIxLTQyNmQtYjVkNy02ZTZkMGJjYTExMjciLCJzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6a3ViZS1zeXN0ZW06Y2x1c3Rlci1hZG1pbiJ9.ZizMo6yKKM3EsBJn9XRjdLYTqaMPQ1h7JFCw26phv35HW2Ew11g_rDwXxsXPvpjs1w8OzVIypGSJhImyU21X3WuiCi8pPHZHrRrQXb2Dy3BplOHnbLmZ6pKzMSODcZixWMJh-9j9CfPpFrxz3vJoz58Sc5SbyHvy9W42qZvL6aRbaMOnPJ0KDf1wixhJDj5T1Mgnb_agRXiT25nA9Ybfw2Tc8kxvE94Hk2Ean3jggYDjiVieGWzvT7bqyyAenFmtuemv1DJvr1qY9LNXQGf6xJfTjPdIvL1FZ1vpoKO4TJnATdYsR--W7gzqVaJZAt6z77z-DXBU2Q04QegLu8c5DXfYNkXaipdS6t-tpt41i4NeVKsiLArkAUehWtmAvbsctRToRobDm2hq6BFOaXvgi3wJWQaq51XfUhMhO6OI2M79eczUnLJwrMb1qf_gDyBqsgeih_rVJkFiB5_BWcje6AMQUvh5GPAKpUFiXf3xfgODvY4kjNnVw5VtP0-NUrYycfX7KcP1nUhPCJcea58WmYcjrCh462b3EQzd1LaKNPmz_5kZMr7KXhaKGGqUg6MyDJpFHh1gZNuWJZf5fyIp10hcA78rPaLtt9UGni-75ciA3JvxyorNvIKpX0xny9zxXHj-df36JS101f46zKVQpFPlB696GzlWLv1S_JX85uk}}]}
```

**Comment by jbuchhammer** [comment]
@worldofgeese : json requires that properties and values are in quotes, e.g. "kind":"Config"
These were obviously lost when writing it to file.

But more important:
Never ever provide your complete kubeconfig in a public repo!
Please remove it immediately from the comment.

**Comment by worldofgeese** [comment]
> @worldofgeese : json requires that properties and values are in quotes, e.g. "kind":"Config"
> These were obviously lost when writing it to file.
> 
> But more important:
> Never ever provide your complete kubeconfig in a public repo!
> Please remove it immediately from the comment.

I haven't. Please review the posted kubeconfig. It's been cleaned of sensitive information.

**Comment by jbuchhammer** [comment]
> 
> 
> > @worldofgeese : json requires that properties and values are in quotes, e.g. "kind":"Config"
> > These were obviously lost when writing it to file.
> > But more important:
> > Never ever provide your complete kubeconfig in a public repo!
> > Please remove it immediately from the comment.
> 
> I haven't. Please review the posted kubeconfig. Sensitive values have been replaced by "stringtoken" and "tokenvalue".

node name: famly-prod-s3hqot7kk3, correct?

**Comment by worldofgeese** [comment]
I did immediately after but I can see GitHub preserves the original comment so I've now deleted the comment.

**Comment by worldofgeese** [comment]
Back to the request at hand, I'd appreciate knowing what the IONOS Cloud team thinks of my proposals. HashiCorp themselves recommends using `local-exec` only as a last resort and I'm inclined to agree.

**Comment by mflorin** [comment]
Hello @worldofgeese ,

Due to security concerns related to storing the kube config in the state, we will not implement a resource specifically for that
nor will we include a `kube_config` attribute in the `ionoscloud_k8s_cluster` resource because the value would end up in the
state.

Our recommendation is to use the `ionoscloud_k8s_cluster` data source which does not store the value of `kube_config` 
which you can extract in your plan, like this:

```
data "ionoscloud_k8s_cluster" "k8s_cluster_example" {
  name     = "k8s-demo"
}

resource "null_resource" "getcfg" {
  provisioner "local-exec" {
    command = "echo \"${data.ionoscloud_k8s_cluster.k8s_cluster_example.kube_config}\" > kubecfg.yaml"
  }
}
```

**Comment by worldofgeese** [comment]
Hi @mflorin,

It would be great if the data resource was usable but it's not usable. The result of echoing the data resource will output to file unusable json. Can you try my code with your data resource? It won't work.

A file object  is a bigger security issue in CI/CD pipelines than a passable data resource. I wrote one proposal in my original post:

**Comment by mflorin** [comment]
@worldofgeese - it cannot work since `config_path` refers to a file system path. `kube_config` from the datasource contains the actual kube_config ... Given that, I'm not sure what you're proposal is exactly.

**Comment by worldofgeese** [comment]
@mflorin the actual kube_config produced is not correct. It cannot be used as it is produced. Can you try and run your examples on a test cluster? Then please try and pass the file to your choice of k9s, Lens or kubectl. They will report that the file is unusable.

**Comment by worldofgeese** [comment]
> Given that, I'm not sure what you're proposal is exactly.

Any form of authentication to the cluster using Terraform constructs that does not rely on what HashiCorp states are last resort methods. Mutable out-of-band files are extremely fragile and vulnerable to accidental git check-ins.

Given this example:
```
provider "k8s" {
  host = rke_cluster.rancher_cluster.api_server_url

  client_certificate     = rke_cluster.rancher_cluster.client_cert
  client_key             = rke_cluster.rancher_cluster.client_key
  cluster_ca_certificate = rke_cluster.rancher_cluster.ca_crt

  load_config_file = false
}
```

could data resources be created to pass in to providers like so?
```
provider "k8s" {
  host = rke_cluster.rancher_cluster.api_server_url

  client_certificate     = data.ionoscloud_k8s_cluster.this.client_cert
  client_key             = data.ionoscloud_k8s_cluster.this.client_key
  cluster_ca_certificate = data.ionoscloud_k8s_cluster.this.ca_crt

  load_config_file = false
}
```

**Comment by mflorin** [comment]
> @mflorin the actual kube_config produced is not correct. It cannot be used as it is produced. Can you try and run your examples on a test cluster? Then please try and pass the file to your choice of k9s, Lens or kubectl. They will report that the file is unusable.

You're right, my example was wrong - it was missing an `yamlencode` - I've edited the comment, please review it.
That said, the datasource works as expected. The actual kube_config **is** correct.

**Comment by mflorin** [comment]
> > Given that, I'm not sure what you're proposal is exactly.
> 
> Any form of authentication to the cluster using Terraform constructs that does not rely on what HashiCorp states are last resort methods. Mutable out-of-band files are extremely fragile and vulnerable to accidental git check-ins. Terraform also has no way of tracking their state. Their outputs cannot be captured in variables, indeed there is little you can do with them inside Terraform data structures.
> 
> Given this example:
> 
> ```
> provider "k8s" {
>   host = rke_cluster.rancher_cluster.api_server_url
> 
>   client_certificate     = rke_cluster.rancher_cluster.client_cert
>   client_key             = rke_cluster.rancher_cluster.client_key
>   cluster_ca_certificate = rke_cluster.rancher_cluster.ca_crt
> 
>   load_config_file = false
> }
> ```
> 
> could data resources be created to pass in to providers like so?
> 
> ```
> provider "k8s" {
>   host = rke_cluster.rancher_cluster.api_server_url
> 
>   client_certificate     = data.ionoscloud_k8s_cluster.this.client_cert
>   client_key             = data.ionoscloud_k8s_cluster.this.client_key
>   cluster_ca_certificate = data.ionoscloud_k8s_cluster.this.ca_crt
> 
>   load_config_file = false
> }
> ```

Yes, we can add new attributes to the `k8s_cluster` data source to reference certificate data.

**Comment by worldofgeese** [comment]
> > @mflorin the actual kube_config produced is not correct. It cannot be used as it is produced. Can you try and run your examples on a test cluster? Then please try and pass the file to your choice of k9s, Lens or kubectl. They will report that the file is unusable.
> 
> You're right, my example was wrong - it was missing an `yamlencode` - I've edited the comment, please review it.
> That said, the datasource works as expected. The actual kube_config **is** correct.

You're right, your datasource does work as expected with the addition of `yamlencode`. You're also right that passing the datasource into `config_path` which expects a file will not work, that was an oversight on my part.

**Comment by worldofgeese** [comment]
> Yes, we can add new attributes to the k8s_cluster data source to reference certificate data.

Thank you very much! That's really great news :confetti_ball: :dancers:

**Comment by mflorin** [comment]
@worldofgeese , would bearer tokens suffice? I'll have to double check but as far as I can see, Ionos Cloud k8s clusters are not configured with client certificate authentication, but with bearer tokens, which I think you could use via the `token` attribute of the hashicorp `kubernetes` resource you're trying to use.

**Comment by worldofgeese** [comment]
@mflorin I've never made use of the `token` attribute before but if you think it will work as a solution, sure!

**Comment by mflorin** [comment]
Hello @worldofgeese,

We've just released `v6.0.0-beta.3` which adds the functionality we discussed.
In short, you can use the user token like this:

```
resource "ionoscloud_k8s_cluster" "test" {
  name = "test_cluster"
  maintenance_window {
    day_of_the_week = "Saturday"
    time            = "03:58:25Z"
  }
}

data "ionoscloud_k8s_cluster" "test" {
  name = "test_cluster"
}

provider "kubernetes" {
  host = data.ionoscloud_k8s_cluster.test.server
  token =  data.ionoscloud_k8s_cluster.test.user_tokens["cluster-admin"]
}
```

We've added 4 new attributes:
- config - a structure that maps the kube_config fields
- user_tokens - a map of authentication tokens for each user
- server - the cluster server
- ca_crt - the cluster ca certificate

For more information about what was added, please see the documentation at https://registry.terraform.io/providers/ionos-cloud/ionoscloud/latest/docs/data-sources/k8s_cluster#config .
